lucene/core/src/java/org/apache/lucene/index/EagerCheapMergePolicy.java

```diff
@@ -0,0 +1,204 @@
+    return cheapMinMergeAtOnce;
+  }
+
+  /** Must have at least this many cheap segments to suggest merging them. */
+  public EagerCheapMergePolicy setCheapMinMergeAtOnce(int cheapMinMergeAtOnce) {
+    this.cheapMinMergeAtOnce = cheapMinMergeAtOnce;
+    return this;
+  }
+
+  /** Returns {@link #setCheapMaxSizeRatio(double)}. */
+  public double getCheapMaxSizeRatio() {
+    return cheapMaxSizeRatio;
+  }
+
+  /**
+   * Limit the maximum ratio of size between a candidate segment (that is deemed cheap) and the sum
+   * of all smaller segments. Example: If it's 2.0, a 20KB segment will merge with a 10KB
+   * segment (or say two 5KB segments) but will not with a 9KB segment.
+   * The limit should be tuned if {@link #setCheapMinMergeAtOnce(int)} is changed. It should be more
+   * than 1/(min-1).
+   */
+  public EagerCheapMergePolicy setCheapMaxSizeRatio(double cheapMaxSizeRatio) {
+    if (cheapMaxSizeRatio <= 0.0) {
+      throw new IllegalArgumentException("must be > 0");
+    }
+    this.cheapMaxSizeRatio = cheapMaxSizeRatio;
+    return this;
+  }
+
+  /** Returns {@link #setCheapLimitConcurrentMergeSegments(int)}. */
+  public int getCheapLimitConcurrentMergeSegments() {
+    return cheapLimitConcurrentMergeSegments;
+  }
+
+  /**
+   * Only find a cheap merge when the number of existing merging segments is less than or equal to
+   * this number.  In all likelihood, those segments are being merged concurrently. By default this
+   * setting is 0, thus won't find cheap merges during concurrent merging. If a higher number is
+   * used (e.g. 10) then there is a likely chance {@link org.apache.lucene.index.ConcurrentMergeScheduler}
+   * will intentionally stall the current thread if it's thresholds have been reached.  This may be
+   * okay if there are multiple merge threads (thus less likely to reach such thresholds) or if it's
+   * deemed more important to do cheap merges at the expense of slowing index throughput further.
+   */
+  public EagerCheapMergePolicy setCheapLimitConcurrentMergeSegments(int cheapLimitConcurrentMergeSegments) {
+    this.cheapLimitConcurrentMergeSegments = cheapLimitConcurrentMergeSegments;
+    return this;
+  }
+
+  @Override
+  public String toString() {
+    return super.toString() +
+        ", cheapMergeThresholdBytes=" + cheapMergeThresholdBytes +
+        ", cheapMaxSizeRatio=" + cheapMaxSizeRatio +
+        ", cheapMinMergeAtOnce=" + cheapMinMergeAtOnce +
+        ", cheapLimitConcurrentMergeSegments=" + cheapLimitConcurrentMergeSegments;
+  }
+
+  @Override
+  public MergeSpecification findFullFlushMerges(MergeTrigger mergeTrigger, SegmentInfos segmentInfos, MergeContext mergeContext) throws IOException {
+    InfoStream infoStream = mergeContext.getInfoStream();
+    boolean verbose = infoStream.isEnabled(INFO_STREAM_CATEGORY);
+
+    if (mergeContext.getMergingSegments().size() <= cheapLimitConcurrentMergeSegments) {
+      OneMerge cheapMerge = findCheapMerge(segmentInfos, mergeContext);
```

Comment:
Hmm, why only one merge?  If enough concurrent threads are indexing, there could be many small segments written during `commit`.